pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test: TestLogic_pg_catalog failed

pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test.TestLogic_pg_catalog [failed](https://mesolite.cluster.engflow.com/invocations/default/75e47700-ffe4-491a-a411-5d1bb5bfb289?testReportRun=2&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy9sb2NhbC1sZWFzZWQtZGVzY3JpcHRvcnM6bG9jYWwtbGVhc2VkLWRlc2NyaXB0b3JzX3Rlc3Q=) on master @ [6816c43040b24bb221d00717c91f6a68a1a8e191](https://github.com/cockroachdb/cockroach/commits/6816c43040b24bb221d00717c91f6a68a1a8e191):

Fatal error:

```
panic: runtime error: slice bounds out of range [:-1]

```
Stack: 

```
goroutine 1129 [running]:
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).processDescriptorUpdate.func1(0xc0057ab6c0, 0x93, 0x8)
	pkg/sql/catalog/lease/lease.go:2191 +0x649
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).processDescriptorUpdate(0xc000f4db08, {0xb4b0bb8, 0xc001c3b980}, 0x93, 0x8, {0x0?, 0x0?})
	pkg/sql/catalog/lease/lease.go:2195 +0x68
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.ensureVersion({0xb4b0bb8, 0xc001c3b980}, 0x93, 0x8, 0xc000f4db08)
	pkg/sql/catalog/lease/lease.go:694 +0x1b1
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).purgeOldVersions(0xc000f4db08, {0xb4b0bb8, 0xc001c3b980}, 0x4da9c9?, 0x93, 0x0, 0x8)
	pkg/sql/catalog/lease/lease.go:1267 +0x18c
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).StartRefreshLeasesTask.func1.2({0xb4b0bb8, 0xc001c3b980})
	pkg/sql/catalog/lease/lease.go:2398 +0x4fd
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).StartRefreshLeasesTask.func1({0xb4b0bb8, 0xc001c3b980})
	pkg/sql/catalog/lease/lease.go:2424 +0x8b4
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0xb4b0bb8, 0xc001c3b980})
	pkg/util/stop/stopper.go:440 +0x112
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 7
	pkg/util/stop/stopper.go:438 +0x247
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic_pg_catalog
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLogic_pg_catalog3080007586
    test_log_scope.go:82: use -show-logs to present logs inline
[13:29:20] setting distsql_workmem='92525B';
[13:29:22] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 14 statements
[13:29:25] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 18 statements
[13:29:41] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 26 statements
[13:29:43] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 39 statements
[13:29:48] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 51 statements
[13:29:50] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 53 statements
[13:29:53] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 56 statements
[13:29:55] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 82 statements
[13:29:57] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 88 statements
    logic.go:3244: let $sourceid = 100130
    logic.go:3244: let $vtableSourceId = 4294967043
[13:29:59] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 105 statements
[13:30:02] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 120 statements
[13:30:05] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 127 statements
[13:30:08] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 130 statements
[13:30:10] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 140 statements
[13:30:12] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 150 statements
[13:30:14] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 172 statements
=== RUN   TestLogic_pg_catalog/31545
[13:30:20] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 195 statements
[13:30:24] --- progress: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog: 198 statements
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_pg_catalog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56988

--- a/pkg/cmd/roachtest/tests/large_schema_benchmark.go
+++ b/pkg/cmd/roachtest/tests/large_schema_benchmark.go
@@ -160,6 +160,9 @@ func registerLargeSchemaBenchmark(r registry.Registry, numTables int, isMultiReg
 						// pg_catalog and information_schema.
 						_, err = conn.Exec("SET CLUSTER SETTING sql.catalog.allow_leased_descriptors.enabled = 'true'")
 						require.NoError(t, err)
+						// Enabled locked descriptor leasing for correctness.
+						_, err = conn.Exec("SET CLUSTER SETTING sql.catalog.descriptor_lease.use_locked_timestamps.enabled = 'true'")
+						require.NoError(t, err)
 						// Since we will be making a large number of databases / tables
 						// quickly,on MR the job retention can slow things down. Let's
 						// minimize how long jobs are kept, so that the creation / ingest

--- a/pkg/sql/catalog/descs/getters.go
+++ b/pkg/sql/catalog/descs/getters.go
@@ -530,6 +530,14 @@ func (b ByIDGetterBuilder) WithoutSynthetic() ByIDGetterBuilder {
 	return b
 }
 
+// WithMetaData configures the byIDGetterBuilder to read zone configs and comments
+// for lease descriptors. This is always acquired for descriptors from storage,
+// and may need an extra round trip.
+func (b ByIDGetterBuilder) WithMetaData() ByIDGetterBuilder {
+	b.flags.layerFilters.withMetadata = true
+	return b
+}
+
 // WithoutDropped configures the ByIDGetterBuilder to error on descriptors
 // which are in a dropped state.
 func (b ByIDGetterBuilder) WithoutDropped() ByIDGetterBuilder {

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -2188,7 +2188,7 @@ func (m *Manager) processDescriptorUpdate(
 			}
 			// Remove one element from the end after.
 			entry.mu.DescriptorIDs = entry.mu.DescriptorIDs[:len(entry.mu.DescriptorIDs)-1]
-			entry.mu.DescriptorVersions = entry.mu.DescriptorVersions[:len(entry.mu.DescriptorIDs)-1]
+			entry.mu.DescriptorVersions = entry.mu.DescriptorVersions[:len(entry.mu.DescriptorVersions)-1]
 			break
 		}
 		return len(entry.mu.DescriptorIDs) == 0

--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -3227,7 +3227,7 @@ CREATE TABLE crdb_internal.create_type_statements (
 				addRow func(...tree.Datum) error,
 			) (matched bool, err error) {
 				id := descpb.ID(tree.MustBeDInt(unwrappedConstraint))
-				scName, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id)
+				scName, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id, true /* includeMetaData */)
 				if err != nil || typDesc == nil {
 					return false, err
 				}

--- a/pkg/sql/logictest/testdata/logic_test/crdb_internal
+++ b/pkg/sql/logictest/testdata/logic_test/crdb_internal
@@ -1,6 +1,6 @@
 # 3node-tenant is blocked from running this file due to heavy reliance on
 # unavailable node IDs in this test.
-# LogicTest: !3node-tenant-default-configs !local-prepared
+# LogicTest: !3node-tenant-default-configs !local-prepared local-leased-descriptors default-configs
 
 query error database "crdb_internal" does not exist
 ALTER DATABASE crdb_internal RENAME TO not_crdb_internal

--- a/pkg/sql/logictest/tests/local-leased-descriptors/BUILD.bazel
+++ b/pkg/sql/logictest/tests/local-leased-descriptors/BUILD.bazel
@@ -12,7 +12,7 @@ go_test(
         "//build/toolchains:is_heavy": {"test.Pool": "heavy"},
         "//conditions:default": {"test.Pool": "large"},
     }),
-    shard_count = 11,
+    shard_count = 12,
     tags = ["cpu:1"],
     deps = [
         "//pkg/base",

--- a/pkg/sql/logictest/tests/local-leased-descriptors/generated_test.go
+++ b/pkg/sql/logictest/tests/local-leased-descriptors/generated_test.go
@@ -80,6 +80,13 @@ func TestLogic_comment_on(
 	runLogicTest(t, "comment_on")
 }
 
+func TestLogic_crdb_internal(
+	t *testing.T,
+) {
+	defer leaktest.AfterTest(t)()
+	runLogicTest(t, "crdb_internal")
+}
+
 func TestLogic_pg_catalog(
 	t *testing.T,
 ) {

--- a/pkg/sql/pg_catalog.go
+++ b/pkg/sql/pg_catalog.go
@@ -3796,9 +3796,13 @@ func addPGAttributeRowForCompositeType(
 }
 
 func getSchemaAndTypeByTypeID(
-	ctx context.Context, p *planner, id descpb.ID,
+	ctx context.Context, p *planner, id descpb.ID, includeMetaData bool,
 ) (catalog.SchemaDescriptor, catalog.TypeDescriptor, error) {
-	typDesc, err := descs.GetCatalogDescriptorGetter(p.Descriptors(), p.txn, &p.EvalContext().Settings.SV).WithoutNonPublic().Get().Type(ctx, id)
+	getter := descs.GetCatalogDescriptorGetter(p.Descriptors(), p.txn, &p.EvalContext().Settings.SV)
+	if includeMetaData {
+		getter = getter.WithMetaData()
+	}
+	typDesc, err := getter.WithoutNonPublic().Get().Type(ctx, id)
 	if err != nil {
 		// If the type was not found, it may be a table.
 		if !(errors.Is(err, catalog.ErrDescriptorNotFound) || pgerror.GetPGCode(err) == pgcode.UndefinedObject) {
@@ -3897,7 +3901,7 @@ https://www.postgresql.org/docs/9.5/catalog-pg-type.html`,
 
 				id := typedesc.UserDefinedTypeOIDToID(ooid)
 
-				sc, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id)
+				sc, typDesc, err := getSchemaAndTypeByTypeID(ctx, p, id, false /* includeMetaData */)
 				if err != nil || typDesc == nil {
 					return false, err
 				}

